src/game/logic/object/weaponset.cpp

```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
+    }
+
+    if (victim->Is_KindOf(KINDOF_MINE) || victim->Is_KindOf(KINDOF_DEMOTRAP)) {
+        return 18;
+    }
+
+    if (!victim->Is_Airborne_Target()) {
+        return 2;
+    }
+
+    if (victim->Is_KindOf(KINDOF_VEHICLE)) {
+        return 1;
+    }
+
+    if (victim->Is_KindOf(KINDOF_INFANTRY)) {
+        return 32;
+    }
+
+    if (victim->Is_KindOf(KINDOF_PARACHUTE)) {
+        return 128;
+    }
+
+    captainslog_dbgassert(victim->Is_KindOf(KINDOF_UNATTACKABLE),
+        "Object %s is being targetted as airborne, but is not infantry, nor vehicle. Is this legit?",
+        victim->Get_Template()->Get_Name().Str());
+
+    return 0;
 }
 
 CanAttackResult WeaponSet::Get_Able_To_Use_Weapon_Against_Target(AbleToAttackType type,
     const Object *source_obj,
-    const Object *obj2,
+    const Object *victim_obj,
     const Coord3D *location,
     CommandSourceType source,
     WeaponSlotType wslot) const
 {
+    int mask;
+
+    if (victim_obj != nullptr) {
+        mask = Get_Victim_Anti_Mask(victim_obj);
+        location = victim_obj->Get_Position();
+    } else {
+        mask = 2;
+    }
+
+    bool is_in_range = false;
+    const Object *containedby = source_obj->Get_Contained_By();
+    bool weapon_in_range = false;
+    bool has_weapon = false;
+
+    for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+        Weapon *weapon = m_weapons[m_curWeapon];
+
+        if (weapon != nullptr) {
+            has_weapon = true;
+
+            if ((mask & m_totalAntiMask) != 0) {
+                bool is_garrison = false;
+                ContainModuleInterface *contain = containedby != nullptr ? containedby->Get_Contain() : nullptr;
+
+                if (contain != nullptr && contain->Is_Garrisonable() && contain->Is_Enclosing_Container_For(source_obj)) {
+                    Coord3D loc = *location;
+                    Coord3D garrison_pos;
+
+                    if (contain->Calc_Best_Garrison_Position(&garrison_pos, &loc)) {
+                        is_in_range = weapon->Is_Source_Object_With_Goal_Position_Within_Attack_Range(
+                            source_obj, &garrison_pos, victim_obj, &loc);
+                        is_garrison = true;
+                    }
+                } else {
+                    is_in_range = victim_obj != nullptr ? weapon->Is_Within_Attack_Range(source_obj, victim_obj) :
+                                                          weapon->Is_Within_Attack_Range(source_obj, location);
+                }
+
+                if (is_in_range) {
+                    weapon_in_range = true;
+                    break;
+                }
+            }
+        }
+    }
+
+    if ((source_obj->Is_KindOf(KINDOF_IMMOBILE) || source_obj->Is_KindOf(KINDOF_SPAWNS_ARE_THE_WEAPONS)
+            || containedby != nullptr)
+        && has_weapon && !weapon_in_range && type != ATTACK_TYPE_4) {
+        return ATTACK_RESULT_1;
+    }
+
+    CanAttackResult result = static_cast<CanAttackResult>(is_in_range + 2);
```

Comment:
Is this a compiler optimization for 
```
CanAttackResult result = is_in_range ? ATTACK_RESULT_3 : ATTACK_RESULT_2
```
?

Reply:
Fixed.